src/coreclr/jit/emitloongarch64.cpp

```diff
@@ -0,0 +1,6431 @@
+ *
+ *  Add an instruction with a single immediate value.
+ */
+
+void emitter::emitIns_I(instruction ins, emitAttr attr, ssize_t imm)
+{
+    code_t code = emitInsCode(ins);
+
+    switch (ins)
+    {
+        case INS_b:
+        case INS_bl:
+            assert(!(imm & 0x3));
+            code |= ((imm >> 18) & 0x3ff);       // offs[25:16]
+            code |= ((imm >> 2) & 0xffff) << 10; // offs[15:0]
+            break;
+        case INS_dbar:
+        case INS_ibar:
+            assert((0 <= imm) && (imm <= 0x7fff));
+            code |= (imm & 0x7fff); // hint
+            break;
+        default:
+            unreached();
+    }
+
+    instrDesc* id = emitNewInstr(attr);
+
+    id->idIns(ins);
+    id->idAddr()->iiaSetInstrEncode(code);
+
+    id->idCodeSize(4);
+    // dispIns(id);
+    appendToCurIG(id);
+}
+
+void emitter::emitIns_I_I(instruction ins, emitAttr attr, ssize_t cc, ssize_t offs)
+{
+#ifdef DEBUG
+    switch (ins)
+    {
+        case INS_bceqz:
+        case INS_bcnez:
+            break;
+        // case INS_:
+        // case INS_:
+        //    break;
+
+        default:
+            unreached();
+    }
+#endif
+
+    code_t code = emitInsCode(ins);
+
+    assert(!(offs & 0x3));
+    assert(!(cc >> 3));
+    code |= ((cc & 0x7) << 5);            // cj
+    code |= ((offs >> 18) & 0x1f);        // offs[20:16]
+    code |= ((offs >> 2) & 0xffff) << 10; // offs[15:0]
+
+    instrDesc* id = emitNewInstr(attr);
+
+    id->idIns(ins);
+    id->idAddr()->iiaSetInstrEncode(code);
+
+    id->idCodeSize(4);
+    // dispIns(id);
+    appendToCurIG(id);
+}
+
+/*****************************************************************************
+ *
+ *  Add an instruction referencing a register and a constant.
+ */
+
+void emitter::emitIns_R_I(instruction ins, emitAttr attr, regNumber reg, ssize_t imm, insOpts opt /* = INS_OPTS_NONE */)
+{
+    code_t code = emitInsCode(ins);
+
+    switch (ins)
+    {
+        case INS_lu12i_w:
+        case INS_lu32i_d:
+        case INS_pcaddi:
+        case INS_pcalau12i:
+        case INS_pcaddu12i:
+        case INS_pcaddu18i:
+            assert(isGeneralRegister(reg));
+            assert((-524288 <= imm) && (imm < 524288));
+
+            code |= reg;                  // rd
+            code |= (imm & 0xfffff) << 5; // si20
+            break;
+        case INS_beqz:
+        case INS_bnez:
+            assert(isGeneralRegisterOrR0(reg));
+            assert(!(imm & 0x3));
+            assert((-1048576 <= (imm >> 2)) && ((imm >> 2) <= 1048575));
+
+            code |= ((imm >> 18) & 0x1f);        // offs[20:16]
+            code |= reg << 5;                    // rj
+            code |= ((imm >> 2) & 0xffff) << 10; // offs[15:0]
+            break;
+        case INS_movfr2cf:
+            assert(isFloatReg(reg));
+            assert((0 <= imm) && (imm <= 7));
+
+            code |= (reg & 0x1f) << 5; // fj
+            code |= imm /*& 0x7*/;     // cc
```

Comment:
Why is the `& 0x7` in a comment? Is there an assumption here that only the low 3 bits are set? Should you add `assert(imm & ~0x7 == 0)`? (same question applies to cases below as well)

Reply:
Here is just for optimization if add there will be more instructions generated by compiler.
In fact these instructions are only used by special cases.

How about adding a assert which only for debug mode ?

Thanks